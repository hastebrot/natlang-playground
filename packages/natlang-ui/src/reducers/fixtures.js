export const paragraphs = `
Ils formaient une assemblée bien grotesque ces êtres singuliers réunis sur le bord de la mare ; les uns avaient leurs plumes tout en désordre, les autres le poil plaqué contre le corps. Tous étaient trempés, de mauvaise humeur, et fort mal à l’aise.

« Comment faire pour nous sécher ? » ce fut la première question, cela va sans dire. Au bout de quelques instants, il sembla tout naturel à Alice de causer familièrement avec ces animaux, comme si elle les connaissait depuis son berceau. Elle eut même une longue discussion avec le Lory, qui, à la fin, lui fit la mine et lui dit d’un air boudeur : « Je suis plus âgé que vous, et je dois par conséquent en savoir plus long. » Alice ne voulut pas accepter cette conclusion avant de savoir l’âge du Lory, et comme celui-ci refusa tout net de le lui dire, cela mit un terme au débat.

Enfin la Souris, qui paraissait avoir un certain ascendant sur les autres, leur cria : « Asseyez-vous tous, et écoutez-moi ! Je vais bientôt vous faire sécher, je vous en réponds ! » Vite, tout le monde s’assit en rond autour de la Souris, sur qui Alice tenait les yeux fixés avec inquiétude, car elle se disait : « Je vais attraper un vilain rhume si je ne sèche pas bientôt. »

« Hum ! » fit la Souris d’un air d’importance ; « êtes-vous prêts ? Je ne sais rien de plus sec que ceci. Silence dans le cercle, je vous prie. « Guillaume le Conquérant, dont le pape avait embrassé le parti, soumit bientôt les Anglais, qui manquaient de chefs, et commençaient à s’accoutumer aux usurpations et aux conquêtes des étrangers. Edwin et Morcar, comtes de Mercie et de Northumbrie — » »
`

export const processParagraphs = paragraphsText => {
  const paragraphs = splitParagraphs(paragraphsText)
  const paragraphsData = paragraphs.map((paragraphText, index) => {
    const words = splitWords(paragraphText)
    const wordsData = words.map((wordText, index) => ({
      type: "word",
      index,
      text: wordText,
      text2: wordText
        .split("")
        .reverse()
        .join(""),
      style: {},
    }))
    return {
      type: "paragraph",
      index,
      text: paragraphText,
      words: wordsData,
      style: {},
    }
  })
  return paragraphsData
}

export const splitParagraphs = text => {
  return text.trim().split(/\n+/)
}

export const splitWords = text => {
  return text.split(/\s+/)
}
