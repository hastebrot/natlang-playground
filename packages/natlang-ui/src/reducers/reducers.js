import { produce } from "immer"
import { processParagraphs, paragraphs } from "./fixtures"

const text = processParagraphs(paragraphs)

export const textInitialState = {
  text: text.slice(0, 1),
}

export const textReducer = (state, action) => {
  if (action.type === "UPDATE_PARAGRAPH") {
    const { paragraph } = action.payload
    return produce(state, state => {
      state.text[paragraph.index].style.color = "red"
      state.text[paragraph.index].style.fontWeight = "bold"
      return state
    })
  } else if (action.type === "UPDATE_WORD") {
    const { paragraph, word } = action.payload
    return produce(state, state => {
      state.text[paragraph.index].words[word.index].style.color = "red"
      state.text[paragraph.index].words[word.index].style.fontWeight = "bold"
      return state
    })
  }
}
