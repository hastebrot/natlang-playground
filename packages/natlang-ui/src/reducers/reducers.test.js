import { useReducer } from "react"
import { renderHook, act, cleanup } from "react-hooks-testing-library"
import { reducer, initialState } from "./reducers"

describe("reducers", () => {
  afterEach(cleanup)

  test("init state", () => {
    // given:
    const { result } = renderHook(() => {
      return useReducer(reducer, initialState)
    })

    // expect:
    const [state] = result.current
    expect(state).toMatchInlineSnapshot(`
Object {
  "count": 0,
}
`)
  })

  test("change state", () => {
    // given:
    const { result } = renderHook(() => {
      return useReducer(reducer, initialState)
    })

    // when:
    act(() => {
      const [state, dispatch] = result.current
      dispatch({ type: "foo" })
    })

    // then:
    const [state] = result.current
    expect(state).toMatchInlineSnapshot(`
Object {
  "count": 1,
}
`)
  })
})
