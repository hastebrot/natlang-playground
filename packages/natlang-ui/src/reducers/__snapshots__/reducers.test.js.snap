// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers process paragraphs 1`] = `
Array [
  Object {
    "index": 0,
    "style": Object {},
    "text": "Ils formaient une assemblée bien grotesque ces êtres singuliers réunis sur le bord de la mare ; les uns avaient leurs plumes tout en désordre, les autres le poil plaqué contre le corps. Tous étaient trempés, de mauvaise humeur, et fort mal à l’aise.",
    "type": "paragraph",
    "words": Array [
      Object {
        "index": 0,
        "style": Object {},
        "text": "Ils",
        "text2": "slI",
        "type": "word",
      },
      Object {
        "index": 1,
        "style": Object {},
        "text": "formaient",
        "text2": "tneiamrof",
        "type": "word",
      },
      Object {
        "index": 2,
        "style": Object {},
        "text": "une",
        "text2": "enu",
        "type": "word",
      },
      Object {
        "index": 3,
        "style": Object {},
        "text": "assemblée",
        "text2": "eélbmessa",
        "type": "word",
      },
      Object {
        "index": 4,
        "style": Object {},
        "text": "bien",
        "text2": "neib",
        "type": "word",
      },
      Object {
        "index": 5,
        "style": Object {},
        "text": "grotesque",
        "text2": "euqsetorg",
        "type": "word",
      },
      Object {
        "index": 6,
        "style": Object {},
        "text": "ces",
        "text2": "sec",
        "type": "word",
      },
      Object {
        "index": 7,
        "style": Object {},
        "text": "êtres",
        "text2": "sertê",
        "type": "word",
      },
      Object {
        "index": 8,
        "style": Object {},
        "text": "singuliers",
        "text2": "sreilugnis",
        "type": "word",
      },
      Object {
        "index": 9,
        "style": Object {},
        "text": "réunis",
        "text2": "sinuér",
        "type": "word",
      },
      Object {
        "index": 10,
        "style": Object {},
        "text": "sur",
        "text2": "rus",
        "type": "word",
      },
      Object {
        "index": 11,
        "style": Object {},
        "text": "le",
        "text2": "el",
        "type": "word",
      },
      Object {
        "index": 12,
        "style": Object {},
        "text": "bord",
        "text2": "drob",
        "type": "word",
      },
      Object {
        "index": 13,
        "style": Object {},
        "text": "de",
        "text2": "ed",
        "type": "word",
      },
      Object {
        "index": 14,
        "style": Object {},
        "text": "la",
        "text2": "al",
        "type": "word",
      },
      Object {
        "index": 15,
        "style": Object {},
        "text": "mare",
        "text2": "eram",
        "type": "word",
      },
      Object {
        "index": 16,
        "style": Object {},
        "text": ";",
        "text2": ";",
        "type": "word",
      },
      Object {
        "index": 17,
        "style": Object {},
        "text": "les",
        "text2": "sel",
        "type": "word",
      },
      Object {
        "index": 18,
        "style": Object {},
        "text": "uns",
        "text2": "snu",
        "type": "word",
      },
      Object {
        "index": 19,
        "style": Object {},
        "text": "avaient",
        "text2": "tneiava",
        "type": "word",
      },
      Object {
        "index": 20,
        "style": Object {},
        "text": "leurs",
        "text2": "sruel",
        "type": "word",
      },
      Object {
        "index": 21,
        "style": Object {},
        "text": "plumes",
        "text2": "semulp",
        "type": "word",
      },
      Object {
        "index": 22,
        "style": Object {},
        "text": "tout",
        "text2": "tuot",
        "type": "word",
      },
      Object {
        "index": 23,
        "style": Object {},
        "text": "en",
        "text2": "ne",
        "type": "word",
      },
      Object {
        "index": 24,
        "style": Object {},
        "text": "désordre,",
        "text2": ",erdroséd",
        "type": "word",
      },
      Object {
        "index": 25,
        "style": Object {},
        "text": "les",
        "text2": "sel",
        "type": "word",
      },
      Object {
        "index": 26,
        "style": Object {},
        "text": "autres",
        "text2": "sertua",
        "type": "word",
      },
      Object {
        "index": 27,
        "style": Object {},
        "text": "le",
        "text2": "el",
        "type": "word",
      },
      Object {
        "index": 28,
        "style": Object {},
        "text": "poil",
        "text2": "liop",
        "type": "word",
      },
      Object {
        "index": 29,
        "style": Object {},
        "text": "plaqué",
        "text2": "éuqalp",
        "type": "word",
      },
      Object {
        "index": 30,
        "style": Object {},
        "text": "contre",
        "text2": "ertnoc",
        "type": "word",
      },
      Object {
        "index": 31,
        "style": Object {},
        "text": "le",
        "text2": "el",
        "type": "word",
      },
      Object {
        "index": 32,
        "style": Object {},
        "text": "corps.",
        "text2": ".sproc",
        "type": "word",
      },
      Object {
        "index": 33,
        "style": Object {},
        "text": "Tous",
        "text2": "suoT",
        "type": "word",
      },
      Object {
        "index": 34,
        "style": Object {},
        "text": "étaient",
        "text2": "tneiaté",
        "type": "word",
      },
      Object {
        "index": 35,
        "style": Object {},
        "text": "trempés,",
        "text2": ",sépmert",
        "type": "word",
      },
      Object {
        "index": 36,
        "style": Object {},
        "text": "de",
        "text2": "ed",
        "type": "word",
      },
      Object {
        "index": 37,
        "style": Object {},
        "text": "mauvaise",
        "text2": "esiavuam",
        "type": "word",
      },
      Object {
        "index": 38,
        "style": Object {},
        "text": "humeur,",
        "text2": ",ruemuh",
        "type": "word",
      },
      Object {
        "index": 39,
        "style": Object {},
        "text": "et",
        "text2": "te",
        "type": "word",
      },
      Object {
        "index": 40,
        "style": Object {},
        "text": "fort",
        "text2": "trof",
        "type": "word",
      },
      Object {
        "index": 41,
        "style": Object {},
        "text": "mal",
        "text2": "lam",
        "type": "word",
      },
      Object {
        "index": 42,
        "style": Object {},
        "text": "à",
        "text2": "à",
        "type": "word",
      },
      Object {
        "index": 43,
        "style": Object {},
        "text": "l’aise.",
        "text2": ".esia’l",
        "type": "word",
      },
    ],
  },
]
`;

exports[`reducers update paragraph style 1`] = `
Object {
  "color": "red",
  "fontWeight": "bold",
}
`;

exports[`reducers update word style 1`] = `
Object {
  "color": "red",
  "fontWeight": "bold",
}
`;
